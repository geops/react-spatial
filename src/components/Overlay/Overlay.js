import React, { Component, useState } from 'react';
import PropTypes from 'prop-types';
import { Resizable } from 're-resizable';
import ResizeHandler from '../ResizeHandler';

const propTypes = {
  /**
   * CSS class added to container.
   */
  className: PropTypes.string,

  /**
   * Children content of the overlay.
   */
  children: PropTypes.node,

  /**
   * Observed element to define screen size.
   */
  observe: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Element),
    PropTypes.instanceOf(Component),
    PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    PropTypes.shape({ current: PropTypes.instanceOf(Component) }),
  ]),

  /**
   * Deactivate the ability to resize the overlay on mobile.
   * /!\ This prop is only used if isMobile (observer width < 768px).
   */
  isMobileResizable: PropTypes.bool,

  /**
   * Size parameters to the Resizable component on mobile.
   * Pass following prop to re-resizable component: 'size', 'defaultSize', 'minHeight' & 'maxHeight'
   * (https://github.com/bokuweb/re-resizable)
   * /!\ This prop is only used if isMobile (observer width < 768px).
   */
  mobileSize: PropTypes.shape({
    minimalHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    maximalHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    defaultSize: PropTypes.shape({
      height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    }),
    size: PropTypes.shape({
      height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    }),
  }),

  /**
   * Minimal width to consider the observed as mobile.
   * Default is 768px.
   */
  thresholdWidthForMobile: PropTypes.number,

  onResizeStop: PropTypes.func,
};

const defaultMobileSize = {
  defaultSize: {
    width: '100%',
    height: '25%',
  },
  size: undefined,
  maximalHeight: '100%',
  minimalHeight: '25%',
};

const defaultProps = {
  className: null,
  children: null,
  observe: null,
  isMobileResizable: true,
  mobileSize: defaultMobileSize,
  thresholdWidthForMobile: 768,
  onResizeStop: () => {},
};

const Overlay = ({
  observe,
  className,
  children,
  isMobileResizable,
  mobileSize,
  onResizeStop,
  thresholdWidthForMobile,
}) => {
  const [isMobile, setIsMobile] = useState();

  const onResize = (entries) => {
    for (let i = 0; i < entries.length; i += 1) {
      const { width } = entries[i].contentRect;
      setIsMobile(width <= thresholdWidthForMobile);
    }
  };

  let resizableMobileSize;
  if (mobileSize) {
    resizableMobileSize = { ...defaultMobileSize, ...mobileSize };
  }

  return (
    <>
      {observe ? <ResizeHandler observe={observe} onResize={onResize} /> : null}
      {isMobile ? (
        <Resizable
          style={{
            position: 'absolute',
          }}
          enable={{
            top: isMobileResizable,
            right: false,
            bottom: false,
            left: false,
            topRight: false,
            bottomRight: false,
            bottomLeft: false,
            topLeft: false,
          }}
          maxHeight={mobileSize && resizableMobileSize.maximalHeight}
          minHeight={mobileSize && resizableMobileSize.minimalHeight}
          handleComponent={{
            top: <div className="tm-overlay-handler">&mdash;</div>,
          }}
          onResizeStop={onResizeStop}
          handleStyles={{
            top: {
              top: '-20px',
              height: '20px',
              padding: '20px 0',
              width: '100%',
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
            },
          }}
          size={resizableMobileSize.size}
          defaultSize={mobileSize && resizableMobileSize.defaultSize}
          className={`tm-overlay-mobile${className ? ` ${className}` : ''}`}
        >
          <div className="tm-overlay-mobile-content">{children}</div>
        </Resizable>
      ) : (
        <div className={`tm-overlay${className ? ` ${className}` : ''}`}>
          {children}
        </div>
      )}
    </>
  );
};

Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;

export default Overlay;
