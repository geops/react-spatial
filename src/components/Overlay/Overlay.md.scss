.tm-overlay-example {
  border: 2px solid #838383;
  background-color: #eee;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  height: 300px;
  width: 100%;

  .tm-overlay-container {
    width: 150px;
  }

  .tm-clickable-feature {
    padding: 10px;
    border-radius: 5px;
    border: 1px solid #838383;
  }

  .tm-overlay-mobile {
    .tm-overlay-mobile-content {
      display: flex;
      justify-content: center;
      align-items: center;

      div {
        width: 80px;
        padding: 10px;
        border-radius: 5px;
        height: unset;
        border: 1px solid #838383;
      }
    }
  }
}
