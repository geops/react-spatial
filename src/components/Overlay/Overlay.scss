.tm-overlay,
.tm-overlay-mobile {
  position: absolute;
  background-color: white;
  left: 0;
  overscroll-behavior: contain;
}

.tm-overlay {
  border-right: 1px solid #eee;
  height: 100%;
}

.tm-overlay-mobile {
  border-top: 1px solid #eee;
  bottom: 0;

  .tm-overlay-mobile-content {
    margin-top: 20px;
    height: calc(100% - 20px);
  }

  .tm-overlay-handler {
    font-weight: bold;
    font-size: 20px;
  }
}
