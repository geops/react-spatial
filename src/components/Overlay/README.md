#

This demonstrates the use of Sidebar.

```jsx
import React, { useState, useEffect, useRef } from 'react';
import Overlay from 'react-spatial/components/Overlay';

function OverlayExample() {
  const [open, setOpen] = useState(true);
  const [ref, setRef] = useState(null);
  const refDiv = useRef(null);

  useEffect(() => {
    setRef(refDiv);
  }, [refDiv]);
  return (
    <div
      className="tm-overlay-example"
      ref={refDiv}
    >
      <div
        role="button"
        className="tm-clickable-feature"
        onClick={() => {
          setOpen(!open);
        }}
      >
        Toggle Overlay
      </div>
      {open && ref ? (
        <Overlay
          observe={ref.current}
          className="tm-overlay-container"
          mobileSize={{
              minimalHeight: '25%',
              maximalHeight: '80%',
              defaultSize: {
                height: '40%',
                width: '100%',
              },
            }}
        >
          <div
            role="button"
            onClick={() => {
              setOpen(false);
            }}
          >
            Close Overlay
          </div>
        </Overlay>
      ) : null}
    </div>
  );
}

<OverlayExample />;
```
