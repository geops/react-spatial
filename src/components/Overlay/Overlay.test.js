import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { act } from 'react-dom/test-utils';
import { configure, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import ResizeObserver from 'resize-observer-polyfill';
import { Resizable } from 're-resizable';
import Overlay from './Overlay';

jest.mock('resize-observer-polyfill');

configure({ adapter: new Adapter() });

const propTypes = {
  isMobileResizable: PropTypes.bool,
  thresholdWidthForMobile: PropTypes.number,
};

const defaultProps = {
  isMobileResizable: undefined,
  thresholdWidthForMobile: undefined,
};

const BasicComponent = ({ thresholdWidthForMobile, isMobileResizable }) => {
  const [ref, setRef] = useState(null);

  return (
    <>
      <div
        ref={(node) => {
          if (node !== ref) {
            setRef(node);
          }
        }}
        className="observer"
      />
      <Overlay
        observe={ref}
        thresholdWidthForMobile={thresholdWidthForMobile}
        isMobileResizable={isMobileResizable}
      >
        Test content
      </Overlay>
    </>
  );
};
BasicComponent.propTypes = propTypes;
BasicComponent.defaultProps = defaultProps;

describe('Overlay', () => {
  test('should match snapshot.', () => {
    const component = renderer.create(<Overlay>Test content</Overlay>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should react on observe resize.', () => {
    const wrapper = mount(<BasicComponent />);
    const target = wrapper.find('.observer').getDOMNode();

    act(() => {
      // The mock class set the onResize property, we just have to run it to
      // simulate a resize
      ResizeObserver.onResize([
        {
          target,
          contentRect: {
            width: 200,
            height: 200,
          },
        },
      ]);
    });
    wrapper.update();

    expect(wrapper.find('.tm-overlay').length > 0).toBe(false);
    expect(wrapper.find('.tm-overlay-mobile').length > 0).toBe(true);
  });

  test('should force mobile overlay display on big screen.', () => {
    const wrapper = mount(
      <BasicComponent thresholdWidthForMobile={Infinity} />,
    );
    const target = wrapper.find('.observer').getDOMNode();

    act(() => {
      ResizeObserver.onResize([
        {
          target,
          contentRect: {
            width: 1200,
            height: 200,
          },
        },
      ]);
    });
    wrapper.update();

    expect(wrapper.find('.tm-overlay').length > 0).toBe(false);
    expect(wrapper.find('.tm-overlay-mobile').length > 0).toBe(true);
  });

  test('should allow resizing with top handler on mobile.', () => {
    const wrapper = mount(<BasicComponent />);
    const target = wrapper.find('.observer').getDOMNode();

    // Force resize to make it mobile.
    act(() => {
      ResizeObserver.onResize([
        {
          target,
          contentRect: {
            width: 200,
            height: 200,
          },
        },
      ]);
    });
    wrapper.update();

    const resizableProps = wrapper.find(Resizable).props();

    expect(resizableProps.enable.top).toBe(true);
  });

  test('should not allow resizing with top handler on mobile.', () => {
    const wrapper = mount(<BasicComponent isMobileResizable={false} />);
    const target = wrapper.find('.observer').getDOMNode();

    // Force resize to make it mobile.
    act(() => {
      ResizeObserver.onResize([
        {
          target,
          contentRect: {
            width: 200,
            height: 200,
          },
        },
      ]);
    });
    wrapper.update();

    const resizableProps = wrapper.find(Resizable).props();

    expect(resizableProps.enable.top).toBe(false);
  });
});
