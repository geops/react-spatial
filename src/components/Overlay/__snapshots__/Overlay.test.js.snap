// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should match snapshot. 1`] = `
<div
  className="tm-overlay"
>
  Test content
</div>
`;
